### PR TITLE
Add support for API paging with limit and continue

### PR DESCRIPTION
**Pull request by agrare**
Implement API chunking support as specified in
https://kubernetes.io/docs/reference/using-api/api-concepts/#retrieving-large-results-sets-in-chunks

Add limit and continue as parameters to get_entities and add continue as an optional attribute (default to nil) to EntityList.

Sample usage:
```
continue_token = nil
begin
  options = {:limit => 100}
  options.merge!(:continue => continue_token) unless continue_token.nil?

  entities = connection.get_pods(options)
  break if entities.nil?

  continue_token = entities.continue

  # do something with the chunk of entities returned
end until entities.continue.nil?
```

Fixes: https://github.com/abonas/kubeclient/issues/283

### CODE REVIEW

**Comment by agrare** [comment]
cc @cben @Ladas

**Comment by abonas** [comment]
> Add limit and continue as parameters to get_entities and get_entities

I guess you meant get and watch? :)

**Comment by agrare** [comment]
> I guess you meant get and watch? :)

Yes thank you!

**Comment by cben** [comment]
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#watch-list-65 :

`continue` — .... This field is not supported if watch is true. ...
`limit` — .... This field is not supported if watch is true. ...

so not relevant for watch.

**Comment by agrare** [comment]
> Watch? What does chunking mean in context of watch? Also, I don't see changes related to watch.

Ah okay they shared the same arguments as get [ref](https://github.com/abonas/kubeclient/blob/061a734c901a1f203c3d8a38bbace01784fc9f00/lib/kubeclient/common.rb#L45) so I thought that was always true.  I'll split those up then so limit/continue aren't possible args for watch

**Comment by agrare** [comment]
@cben fixed

**Comment by agrare** [comment]
@cben as for the higher-level vs lower-level interface, I actually like being able to process the chunks rather than just getting an iterator because I want to save the smaller chunks as well.

I can definitely see the benefit of a "nicer" iterator though where the caller doesn't have to worry about it, but this is closer to how the API is implemented.

**Comment by agrare** [comment]
@cben I pushed a version which always uses paging and handles the continue internally, either yielding the batches or returning the full list if no block is given.  This makes it easier on the callers but I haven't found a nice way of cleanly handling the format_response since it parses the body and I had to parse the JSON to get the continue out of the payload.

File: lib/kubeclient/entity_list.rb [hunk 1]
```diff
@@ -5,12 +5,17 @@ module Common
...
-      def initialize(kind, resource_version, list)
+      def initialize(kind, resource_version, list, continue = nil)
         @kind = kind
         # rubocop:disable Style/VariableName
         @resourceVersion = resource_version
+        @continue = continue
         super(list)
       end
+
+      def last?
```
Comment (cben) [inline lib/kubeclient/entity_list.rb]: Nice!  I think this is worth keeping if we return to low-level version.

Comment (agrare) [reply lib/kubeclient/entity_list.rb]: Yeah absolutely, this made the calling code a lot cleaner

**Comment by agrare** [comment]
> I now lean towards the low-level version, which does the job, is 100% opt-in, and has no open questions.

Yeah I agree, it is simple and lets us get this functionality in without changing the default behavior.
I'll work on the docs/changelog/tests now, thanks @cben !

**Comment by agrare** [comment]
> docs (with usage example; mention risk of 410 Gone; mention servers may ignore it). changelog if possible.

Added, please check that they meet with what you expected.  I added an Unreleased section to the changelog similar to how we handle per-sprint manageiq changelogs, when you cut another release you can just replace that with the release number.

> some tests? I'd include both responses from server supporting limit/continue, and a server that ignores limit.

I added a simple test where two pages are returned.  I am going to add a test for 410 Gone condition as well.  I'll need to try to find a pre-1.9 cluster to test a server that doesn't support chunking though, any hints?

File: README.md [hunk 2]
```diff
@@ -345,6 +345,29 @@ You can specify multiple labels (that option will return entities which have bot
...
+Get all entities of a specific type in chunks:<br>
+
+```ruby
+continue = nil
+options = {
+  limit: 1_000
+}
+
+begin
+  options[:continue] = continue unless continue.nil?
```
Comment (cben) [inline README.md]: I propose documenting that user may always pass `continue: nil` with same result as omitting it.
Then user could do:
```
continue = nil
begin
  entities = client.get_pods(limit: 1_000, continue: continue)

  continue = entities.continue
end until entities.last?
```

P.S. googling this [`begin ... end until ...` construct](https://rosettacode.org/wiki/Loops/Do-while#Ruby), turns out Matz [regrets it and considers removing it](http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/6745).  So let's write this as `loop` with `break if entities.last?` inside.

File: README.md [hunk 3]
```diff
@@ -345,6 +345,29 @@ You can specify multiple labels (that option will return entities which have bot
 pods = client.get_pods(label_selector: 'name=redis-master,app=redis')
 ```
 
+Get all entities of a specific type in chunks:<br>
```
Comment (cben) [inline README.md]: nit: the `<br>` is unnecessary, especially as it's followed by empty line.

File: README.md (hunk 2 above)
Comment (agrare) [reply README.md]: :+1: it looks like kubernetes does handle being passed a nil continue option on the first get so yeah this is much cleaner

Comment (agrare) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (agrare) [reply README.md]: Done

**Comment by agrare** [comment]
@cben can you take another look?  I updated the readme and added tests for 410 gone

File: README.md (hunk 2 above)
Comment (cben) [reply README.md]: kubernetes doesn't see the nil, it's dropped by
```
SEARCH_ARGUMENTS.each { |k, v| params[k] = options[v] if options[v] }
```

**Comment by cben** [comment]
LGTM :100:

**Comment by agrare** [comment]
Thanks so much @cben!

**Pull request merged**